File: py/arolla/operators/standard/bool.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Declaration of M.bool.* operators."""

from arolla import arolla
from arolla.operators.standard import edge as M_edge
from arolla.operators.standard import qtype as M_qtype

M = arolla.M
P = arolla.P
constraints = arolla.optools.constraints


_like_binary_compare = {
    'qtype_constraints': [
        constraints.has_scalar_type(P.x),
        constraints.has_scalar_type(P.y),
        constraints.expect_implicit_cast_compatible(P.x, P.y),
    ],
    'qtype_inference_expr': M_qtype.broadcast_qtype_like(
        M_qtype.broadcast_qtype_like(P.x, P.y), arolla.BOOLEAN
    ),
}


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.equal', **_like_binary_compare)
def equal(x, y):
  """Equal to.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if the arguments are equal. Returns False if the arguments are
    not equal. Missing if either of the arguments is missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.less_equal', **_like_binary_compare)
def less_equal(x, y):
  """Less than or Equal to.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if the first argument is less than or equal to the second.
    Returns False if the first argument is greater than the second. Missing if
    either of the arguments is missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.less', **_like_binary_compare)
def less(x, y):
  """Less than.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if the first argument is less than the second. Returns False if
    the first argument is greater than or equal to the second. Missing if either
    of the arguments is missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_lambda_operator('bool.greater_equal')
def greater_equal(x, y):
  """Greater than or Equal to.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if the first argument is greater than or equal to the second.
    Returns False if the first argument is less than the second. Missing if
    either of the arguments is missing.
  """
  return less_equal(y, x)  # pylint: disable=arguments-out-of-order


@arolla.optools.add_to_registry()
@arolla.optools.as_lambda_operator('bool.greater')
def greater(x, y):
  """Greater than.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if the first argument is greater than the second. Returns False
    if the first argument is leass than or equal to the second. Missing if
    either of the arguments is missing.
  """
  return less(y, x)  # pylint: disable=arguments-out-of-order


_like_binary_logical = {
    'qtype_constraints': [
        constraints.expect_booleans(P.x),
        constraints.expect_booleans(P.y),
        constraints.expect_implicit_cast_compatible(P.x, P.y),
    ],
    'qtype_inference_expr': M_qtype.common_qtype(P.x, P.y),
}


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.logical_and', **_like_binary_logical)
def logical_and(x, y):
  """Ternary And.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    True if all args are True. Returns False if at least one arg is
    False. Otherwise missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.logical_or', **_like_binary_logical)
def logical_or(x, y):
  """Ternary Or.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    False if all args are False. Returns True if at least one arg is
    True. Otherwise missing.
  """
  raise NotImplementedError('provided by backend')


_like_unary_logical = {
    'qtype_constraints': [constraints.expect_booleans(P.x)],
    'qtype_inference_expr': P.x,
}


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.logical_not', **_like_unary_logical)
def logical_not(x):
  """Ternary Not.

  Args:
    x: First argument.

  Returns:
    False if the argument is True. Returns True if the argument is False.
    Otherwise missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator('bool.not_equal', **_like_binary_compare)
def not_equal(x, y):
  """Ternary Not Equal to.

  Args:
    x: First argument.
    y: Second argument.

  Returns:
    False if the arguments are equal. Returns True if the arguments are not
    equal. Missing if either of the arguments is missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator(
    'bool.logical_if',
    qtype_constraints=[
        constraints.expect_booleans(P.condition),
        constraints.expect_implicit_cast_compatible(
            P.true_value, P.false_value, P.missing_value
        ),
        constraints.expect_broadcast_compatible(
            P.condition, P.true_value, P.false_value, P.missing_value
        ),
    ],
    qtype_inference_expr=M_qtype.conditional_qtype(
        P.condition == arolla.OPTIONAL_BOOLEAN,
        constraints.common_qtype_expr(
            P.true_value, P.false_value, P.missing_value
        ),
        M_qtype.broadcast_qtype_like(
            P.condition,
            constraints.common_qtype_expr(
                P.true_value, P.false_value, P.missing_value
            ),
        ),
    ),
)
def logical_if(condition, true_value, false_value, missing_value):
  """Ternary If.

  Args:
    condition: Boolean condition.
    true_value: Value corresponding to the True condition.
    false_value: Value corresponding to the False condition.
    missing_value: Value corresponding to the missing condition value.

  Returns:
    The second/third/fourth argument, depending on on whether the first argument
    is True, False, or missing.
  """
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator(
    'bool._logical_all',
    qtype_constraints=[
        constraints.expect_array(P.x),
        constraints.expect_booleans(P.x),
        *constraints.expect_edge(P.into, child_side_param=P.x),
    ],
    qtype_inference_expr=M_qtype.with_value_qtype(
        M_qtype.get_parent_shape_qtype(P.into), arolla.BOOLEAN
    ),
)
def _logical_all(x, into):
  """(internal) Aggregates values by `bool.logical_and` along an edge."""
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_lambda_operator(
    'bool.logical_all',
    qtype_constraints=[
        constraints.expect_array(P.x),
        constraints.expect_booleans(P.x),
        *constraints.expect_edge_or_unspecified(P.into, child_side_param=P.x),
    ],
)
def logical_all(x, into=arolla.unspecified()):
  """Aggregates values by `bool.logical_and` along an edge."""
  return _logical_all(
      x, M.core.default_if_unspecified(into, M_edge.to_scalar(x))
  )


@arolla.optools.add_to_registry()
@arolla.optools.as_backend_operator(
    'bool._logical_any',
    qtype_constraints=[
        constraints.expect_array(P.x),
        constraints.expect_booleans(P.x),
        *constraints.expect_edge(P.into, child_side_param=P.x),
    ],
    qtype_inference_expr=M_qtype.with_value_qtype(
        M_qtype.get_parent_shape_qtype(P.into), arolla.BOOLEAN
    ),
)
def _logical_any(x, into):
  """(internal) Aggregates values by `bool.logical_or` along an edge."""
  raise NotImplementedError('provided by backend')


@arolla.optools.add_to_registry()
@arolla.optools.as_lambda_operator(
    'bool.logical_any',
    qtype_constraints=[
        constraints.expect_array(P.x),
        constraints.expect_booleans(P.x),
        *constraints.expect_edge_or_unspecified(P.into, child_side_param=P.x),
    ],
)
def logical_any(x, into=arolla.unspecified()):
  """Aggregates values by `bool.logical_or` along an edge."""
  return _logical_any(
      x, M.core.default_if_unspecified(into, M_edge.to_scalar(x))
  )

File: arolla/expr/optimization/peephole_optimizations/bool.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include "arolla/expr/optimization/peephole_optimizations/bool.h"

#include <array>
#include <functional>
#include <string>
#include <utility>
#include <vector>

#include "absl/container/flat_hash_set.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "arolla/expr/expr.h"
#include "arolla/expr/expr_node.h"
#include "arolla/expr/optimization/peephole_optimizer.h"
#include "arolla/expr/registered_expr_operator.h"
#include "arolla/memory/optional_value.h"
#include "arolla/qtype/qtype_traits.h"
#include "arolla/util/fingerprint.h"
#include "arolla/util/status_macros_backport.h"

namespace arolla::expr {

namespace {

auto Matches(const std::vector<ExprNodePtr>& patterns) {
  absl::flat_hash_set<Fingerprint> pattern_prints;
  pattern_prints.reserve(patterns.size());
  for (const auto& p : patterns) {
    pattern_prints.insert(p->fingerprint());
  }
  return [pattern_prints(std::move(pattern_prints))](const ExprNodePtr& node) {
    return pattern_prints.contains(node->fingerprint());
  };
}

std::vector<ExprNodePtr> BoolLiterals(bool value) {
  return {Literal(value), Literal(MakeOptionalValue(value))};
}

constexpr std::array kComparisonOppositeOps = {
    std::pair{"bool.equal", "bool.not_equal"},
    std::pair{"bool.not_equal", "bool.equal"},
    // bool.greater and bool.greater_equal are not backend operators,
    // so we don't transform from them.
    std::pair{"bool.less", "bool.greater_equal"},
    std::pair{"bool.less_equal", "bool.greater"}};

// Remove `bool.logical_not`:
// 1. double negation
// 2. around comparison operation by replacing operation with an opposite one.
absl::Status LogicalNotComparisonOptimizations(
    PeepholeOptimizationPack& optimizations) {
  ExprNodePtr a = Placeholder("a");
  ExprNodePtr b = Placeholder("b");
  {  // double negation
    ASSIGN_OR_RETURN(
        ExprNodePtr from,
        CallOpReference("bool.logical_not",
                        {CallOpReference("bool.logical_not", {a})}));
    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(from, a));
  }
  for (auto [cmp1, cmp2] : kComparisonOppositeOps) {
    ASSIGN_OR_RETURN(
        ExprNodePtr from,
        CallOpReference("bool.logical_not", {CallOpReference(cmp1, {a, b})}));
    ASSIGN_OR_RETURN(ExprNodePtr to, CallOpReference(cmp2, {a, b}));
    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(from, to));
  }
  return absl::OkStatus();
}

constexpr std::array kComparisonOps = {"equal", "not_equal", "less",
                                       "less_equal"};
constexpr std::array kLogicalOps = {"and", "or"};

// Optimization to collapse
// `core.equal(True, a)` -> `core.equal(a, True)`
//
// `core.equal(bool.{comparison_op}(a, b), True)` ->
// `core.{comparison_op}(a, b)`
//
// `core.equal(core.to_optional._scalar(bool.{comparison_op}(a, b)), True)` ->
// `core.{comparison_op}(a, b)`
//
// `core.equal(bool.logical_{and,or}(a, b), True)` ->
// `core.presence_{and,or}(core.equal(a, True), core.equal(b, True))`
// when one or both of a and b are either literal or bool comparison operation.
//
// Such patterns are often happen during automatic conversion from 3-bool
// to 2-bool logic.
absl::Status CoreBoolComparisonOptimizations(
    PeepholeOptimizationPack& optimizations) {
  ExprNodePtr a = Placeholder("a");
  ExprNodePtr b = Placeholder("b");
  ExprNodePtr c = Placeholder("c");
  ExprNodePtr d = Placeholder("d");
  ExprNodePtr true_ = Placeholder("true");
  std::vector<ExprNodePtr> true_literals = BoolLiterals(true);
  auto is_true = Matches(true_literals);

  {  // true == a -> a == true in order to reduce number of optimizations
    ASSIGN_OR_RETURN(ExprNodePtr from,
                     CallOpReference("core.equal", {true_, a}));
    ASSIGN_OR_RETURN(ExprNodePtr to, CallOpReference("core.equal", {a, true_}));
    ASSIGN_OR_RETURN(
        optimizations.emplace_back(),
        PeepholeOptimization::CreatePatternOptimization(
            from, to, {{"true", is_true}, {"a", std::not_fn(is_true)}}));
  }
  for (absl::string_view comparison_op : kComparisonOps) {
    ASSIGN_OR_RETURN(
        ExprNodePtr bool_cmp,
        CallOpReference(absl::StrCat("bool.", comparison_op), {a, b}));
    ASSIGN_OR_RETURN(
        ExprNodePtr core_cmp,
        CallOpReference(absl::StrCat("core.", comparison_op), {a, b}));
    {
      ASSIGN_OR_RETURN(ExprNodePtr from,
                       CallOpReference("core.equal", {bool_cmp, true_}));
      ASSIGN_OR_RETURN(optimizations.emplace_back(),
                       PeepholeOptimization::CreatePatternOptimization(
                           from, core_cmp, {{"true", is_true}}));
    }
    {
      ASSIGN_OR_RETURN(
          ExprNodePtr from,
          CallOpReference(
              "core.equal",
              {CallOpReference("core.to_optional._scalar", {bool_cmp}),
               true_}));
      ASSIGN_OR_RETURN(optimizations.emplace_back(),
                       PeepholeOptimization::CreatePatternOptimization(
                           from, core_cmp, {{"true", is_true}}));
    }
  }

  absl::flat_hash_set<std::string> bool_comparison_ops;
  for (absl::string_view comparison_op : kComparisonOps) {
    bool_comparison_ops.insert(absl::StrCat("bool.", comparison_op));
  }
  auto eq_true_will_be_optimized_further =
      [bool_comparison_ops](const ExprNodePtr& node) {
        if (node->is_literal()) return true;
        if (!node->is_op()) return false;
        return IsRegisteredOperator(node->op()) &&
               bool_comparison_ops.contains(node->op()->display_name());
      };
  for (absl::string_view logical_op : kLogicalOps) {
    ASSIGN_OR_RETURN(
        ExprNodePtr bool_logic,
        CallOpReference(absl::StrCat("bool.logical_", logical_op), {a, b}));
    ASSIGN_OR_RETURN(
        ExprNodePtr core_logic,
        CallOpReference(absl::StrCat("core.presence_", logical_op),
                        {CallOpReference("core.equal", {a, true_}),
                         CallOpReference("core.equal", {b, true_})}));
    {
      ASSIGN_OR_RETURN(ExprNodePtr from,
                       CallOpReference("core.equal", {bool_logic, true_}));

      ASSIGN_OR_RETURN(
          optimizations.emplace_back(),
          PeepholeOptimization::CreatePatternOptimization(
              from, core_logic,
              {{"true", is_true}, {"a", eq_true_will_be_optimized_further}}));
      ASSIGN_OR_RETURN(
          optimizations.emplace_back(),
          PeepholeOptimization::CreatePatternOptimization(
              from, core_logic,
              {{"true", is_true}, {"b", eq_true_will_be_optimized_further}}));
    }
  }

  return absl::OkStatus();
}

// Optimization to remove unused branches in `bool.logical_if`.
absl::Status LogicalIfOptimizations(PeepholeOptimizationPack& optimizations) {
  ExprNodePtr condition = Placeholder("condition");
  ExprNodePtr a = Placeholder("a");
  ExprNodePtr b = Placeholder("b");
  ExprNodePtr c = Placeholder("c");
  auto is_scalar_bool = [](const ExprNodePtr& expr) {
    return expr->qtype() == GetQType<bool>();
  };
  ExprNodePtr true_ = Placeholder("true");
  std::vector<ExprNodePtr> true_literals = BoolLiterals(true);
  auto is_true = Matches(true_literals);

  ExprNodePtr false_ = Placeholder("false");
  std::vector<ExprNodePtr> false_literals = BoolLiterals(false);
  auto is_false = Matches(false_literals);
  // Case when cond is never missing or missing interpreted as false.
  // bool.logical_if(core.to_optional._scalar(cond), a, b, c)
  // AND
  // bool.logical_if(cond | false, a, b, c)
  // ->
  // core.where(cond == true, a, b)
  {
    ASSIGN_OR_RETURN(
        ExprNodePtr from1,
        CallOpReference(
            "bool.logical_if",
            {CallOpReference("core.to_optional._scalar", {condition}), a, b,
             c}));
    ASSIGN_OR_RETURN(
        ExprNodePtr to,
        CallOpReference(
            "core.where",
            {CallOpReference("core.equal", {condition, Literal(true)}), a, b}));

    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(
                         from1, to, {{"condition", is_scalar_bool}}));

    ASSIGN_OR_RETURN(ExprNodePtr from2,
                     CallOpReference("bool.logical_if",
                                     {CallOpReference("core.presence_or",
                                                      {condition, false_}),
                                      a, b, c}));
    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(
                         from2, to, {{"false", is_false}}));
  }
  // Case when false and missing cases are identical.
  // bool.logical_if(cond, a, b, b) ->
  // core.where(cond == true, a, b)
  // Here we rely on the CoreBoolComparisonOptimizations that will happen
  // downstream and likely remove the `cond == true` part.
  {
    ASSIGN_OR_RETURN(ExprNodePtr from,
                     CallOpReference("bool.logical_if", {condition, a, b, b}));
    ASSIGN_OR_RETURN(
        ExprNodePtr to,
        CallOpReference(
            "core.where",
            {CallOpReference("core.equal", {condition, Literal(true)}), a, b}));
    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(from, to));
  }
  // Case when missing is interpreted as true.
  // bool.logical_if(cond | true, a, b, c) ->
  // core.where(cond == false, b, a)
  {
    ASSIGN_OR_RETURN(
        ExprNodePtr from,
        CallOpReference("bool.logical_if", {CallOpReference("core.presence_or",
                                                            {condition, true_}),
                                            a, b, c}));
    ASSIGN_OR_RETURN(
        ExprNodePtr to,
        CallOpReference(
            "core.where",
            {CallOpReference("core.equal", {condition, Literal(false)}), b,
             a}));

    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(
                         from, to, {{"true", is_true}}));
  }
  // Case when the 2 arguments after cond are true and false.
  // bool.logical_if(cond, true, false, a) ->
  // core.presence_or(cond, a)
  {
    ASSIGN_OR_RETURN(
        ExprNodePtr from,
        CallOpReference("bool.logical_if", {condition, true_, false_, a}));
    ASSIGN_OR_RETURN(ExprNodePtr to,
                     CallOpReference("core.presence_or", {condition, a}));

    ASSIGN_OR_RETURN(optimizations.emplace_back(),
                     PeepholeOptimization::CreatePatternOptimization(
                         from, to, {{"true", is_true}, {"false", is_false}}));
  }
  return absl::OkStatus();
}

}  // namespace

absl::StatusOr<PeepholeOptimizationPack> BoolOptimizations() {
  PeepholeOptimizationPack optimizations;
  RETURN_IF_ERROR(LogicalNotComparisonOptimizations(optimizations));
  RETURN_IF_ERROR(CoreBoolComparisonOptimizations(optimizations));
  RETURN_IF_ERROR(LogicalIfOptimizations(optimizations));
  return optimizations;
}

}  // namespace arolla::expr
